### ISSUE TITLE
Unable to build osgEarthUtils: writeSQIDfile unresolved symbol (see below, solution is related to missing Ws2_32.lib)

### ISSUE THREAD

**Issue by beachwalker** [gwaldron/osgearth#995]
Unresolved symbol when trying to build on Windows x64:
MGRSGraticule.obj : error LNK2019: Verweis auf nicht aufgel÷stes externes Symbo
l "__imp_htonl" in Funktion ""void __cdecl `anonymous namespace'::writeSQIDfile
(class osgEarth::URI const &)" (?writeSQIDfile@?A0x06592c5b@@YAXAEBVURI@osgEart
h@@@Z)". [I:\Projects\OsgEarth\build\x64\Release\osgearth\src\osgEarthUtil\osgE
arthUtil.vcxproj]
MGRSGraticule.obj : error LNK2019: Verweis auf nicht aufgel÷stes externes Symbo
l "__imp_htons" in Funktion ""void __cdecl `anonymous namespace'::writeSQIDfile
(class osgEarth::URI const &)" (?writeSQIDfile@?A0x06592c5b@@YAXAEBVURI@osgEart
h@@@Z)". [I:\Projects\OsgEarth\build\x64\Release\osgearth\src\osgEarthUtil\osgE
arthUtil.vcxproj]
MGRSGraticule.obj : error LNK2019: Verweis auf nicht aufgel÷stes externes Symbo
l "__imp_ntohl" in Funktion ""bool __cdecl `anonymous namespace'::readSQIDfile(
class osgEarth::URI const &,class std::list<class osg::ref_ptr<class osgEarth::
Features::Feature>,class std::allocator<class osg::ref_ptr<class osgEarth::Feat
ures::Feature> > > &)" (?readSQIDfile@?A0x06592c5b@@YA_NAEBVURI@osgEarth@@AEAV?
$list@V?$ref_ptr@VFeature@Features@osgEarth@@@osg@@V?$allocator@V?$ref_ptr@VFea
ture@Features@osgEarth@@@osg@@@std@@@std@@@Z)". [I:\Projects\OsgEarth\build\x64
\Release\osgearth\src\osgEarthUtil\osgEarthUtil.vcxproj]
MGRSGraticule.obj : error LNK2019: Verweis auf nicht aufgel÷stes externes Symbo
l "__imp_ntohs" in Funktion ""bool __cdecl `anonymous namespace'::readSQIDfile(
class osgEarth::URI const &,class std::list<class osg::ref_ptr<class osgEarth::
Features::Feature>,class std::allocator<class osg::ref_ptr<class osgEarth::Feat
ures::Feature> > > &)" (?readSQIDfile@?A0x06592c5b@@YA_NAEBVURI@osgEarth@@AEAV?
$list@V?$ref_ptr@VFeature@Features@osgEarth@@@osg@@V?$allocator@V?$ref_ptr@VFea
ture@Features@osgEarth@@@osg@@@std@@@std@@@Z)". [I:\Projects\OsgEarth\build\x64
\Release\osgearth\src\osgEarthUtil\osgEarthUtil.vcxproj]
I:\Projects\OsgEarth\build\x64\Release\osgearth\lib\Release\osgEarthUtil.dll :
fatal error LNK1120: 4 nicht aufgel÷ste Externe [I:\Projects\OsgEarth\build\x64
\Release\osgearth\src\osgEarthUtil\osgEarthUtil.vcxproj]

"I:\Projects\OsgEarth\build\x64\Release\osgearth\install.vcxproj" (Standardziel
) (1) ->
"I:\Projects\OsgEarth\build\x64\Release\osgearth\ALL_BUILD.vcxproj" (Standardzi
el) (3) ->
"I:\Projects\OsgEarth\build\x64\Release\osgearth\src\applications\osgearth_3pv\
application_osgearth_3pv.vcxproj" (Standardziel) (4) ->
"I:\Projects\OsgEarth\build\x64\Release\osgearth\src\osgEarthUtil\osgEarthUtil.
vcxproj" (Standardziel) (9) ->
(Link Ziel) ->
  MGRSGraticule.obj : error LNK2019: Verweis auf nicht aufgel÷stes externes Sym
bol "__imp_htonl" in Funktion ""void __cdecl `anonymous namespace'::writeSQIDfi
le(class osgEarth::URI const &)" (?writeSQIDfile@?A0x06592c5b@@YAXAEBVURI@osgEa
rth@@@Z)". [I:\Projects\OsgEarth\build\x64\Release\osgearth\src\osgEarthUtil\os
gEarthUtil.vcxproj]
  MGRSGraticule.obj : error LNK2019: Verweis auf nicht aufgel÷stes externes Sym
bol "__imp_htons" in Funktion ""void __cdecl `anonymous namespace'::writeSQIDfi
le(class osgEarth::URI const &)" (?writeSQIDfile@?A0x06592c5b@@YAXAEBVURI@osgEa
rth@@@Z)". [I:\Projects\OsgEarth\build\x64\Release\osgearth\src\osgEarthUtil\os
gEarthUtil.vcxproj]
  MGRSGraticule.obj : error LNK2019: Verweis auf nicht aufgel÷stes externes Sym
bol "__imp_ntohl" in Funktion ""bool __cdecl `anonymous namespace'::readSQIDfil
e(class osgEarth::URI const &,class std::list<class osg::ref_ptr<class osgEarth
::Features::Feature>,class std::allocator<class osg::ref_ptr<class osgEarth::Fe
atures::Feature> > > &)" (?readSQIDfile@?A0x06592c5b@@YA_NAEBVURI@osgEarth@@AEA
V?$list@V?$ref_ptr@VFeature@Features@osgEarth@@@osg@@V?$allocator@V?$ref_ptr@VF
eature@Features@osgEarth@@@osg@@@std@@@std@@@Z)". [I:\Projects\OsgEarth\build\x
64\Release\osgearth\src\osgEarthUtil\osgEarthUtil.vcxproj]
  MGRSGraticule.obj : error LNK2019: Verweis auf nicht aufgel÷stes externes Sym
bol "__imp_ntohs" in Funktion ""bool __cdecl `anonymous namespace'::readSQIDfil
e(class osgEarth::URI const &,class std::list<class osg::ref_ptr<class osgEarth
::Features::Feature>,class std::allocator<class osg::ref_ptr<class osgEarth::Fe
atures::Feature> > > &)" (?readSQIDfile@?A0x06592c5b@@YA_NAEBVURI@osgEarth@@AEA
V?$list@V?$ref_ptr@VFeature@Features@osgEarth@@@osg@@V?$allocator@V?$ref_ptr@VF
eature@Features@osgEarth@@@osg@@@std@@@std@@@Z)". [I:\Projects\OsgEarth\build\x
64\Release\osgearth\src\osgEarthUtil\osgEarthUtil.vcxproj]
  I:\Projects\OsgEarth\build\x64\Release\osgearth\lib\Release\osgEarthUtil.dll
: fatal error LNK1120: 4 nicht aufgel÷ste Externe [I:\Projects\OsgEarth\build\x
64\Release\osgearth\src\osgEarthUtil\osgEarthUtil.vcxproj]

**Comment by Aea** [comment]
Well this is a curious thread to get highlighted on.

**Comment by beachwalker** [comment]
@Aea Your name is too short and common. ;-) Any idea how to reformat to suppress this?

**Comment by beachwalker** [comment]
My CMake settings for completeness:

```
set(CMAKE_INSTALL_PREFIX "../../../../install/x64/Release/osgearth" CACHE PATH "")
set(CMAKE_CONFIGURATION_TYPES "Release" CACHE STRING "")
set(INSTALL_BIN_DIR "bin" CACHE PATH "Installation directory for executables")
set(INSTALL_LIB_DIR "lib" CACHE PATH "Installation directory for libraries")
set(INSTALL_INC_DIR "include" CACHE PATH "Installation directory for headers")
set(INSTALL_MAN_DIR "share/man" CACHE PATH "Installation directory for manual pages")
set(INSTALL_PKGCONFIG_DIR "share/pkgconfig" CACHE PATH "Installation directory for pkgconfig (.pc) files")
set(WIN32_USE_MP 1 CACHE BOOL "")
set(APPEND_OPENSCENEGRAPH_VERSION 0 CACHE BOOL "")
set(OSG_DIR "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/osg" CACHE PATH "")
set(OSG_VERSION_EXE "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/osg/bin/osgversion.exe" CACHE FILEPATH "")
set(ZLIB_ROOT "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/zlib" CACHE PATH "Zlib root")
#set(ZLIB_LIBRARY "${CMAKE_SOURCE_DIR}/../../../../install/zlib/lib/zlib.lib" CACHE FILEPATH "")
set(CURL_INCLUDE_DIR "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/curl/include" CACHE PATH "")
set(LIBNOISE_INCLUDE_DIR "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/noise/include" CACHE PATH "")
set(LIBNOISE_LIBRARY "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/noise/lib/libnoise.lib" CACHE FILEPATH "")
set(LIBNOISE_LIBRARY_DEBUG "${CMAKE_SOURCE_DIR}/../../../../install/x64/Debug/noise/lib/libnoised.lib" CACHE FILEPATH "")
set(CURL_IS_STATIC 0 CACHE BOOL "")
set(CURL_LIBRARY "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/curl/lib/libcurl_imp.lib" CACHE FILEPATH "")
set(CURL_LIBRARY_DEBUG "${CMAKE_SOURCE_DIR}/../../../../install/x64/Debug/curl/lib/libcurld_imp.lib" CACHE FILEPATH "")
set(GEOS_INCLUDE_DIR "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/geos/include" CACHE PATH "")
set(GEOS_LIBRARY "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/geos/lib/geos.lib" CACHE FILEPATH "")
set(GEOS_LIBRARY_DEBUG "${CMAKE_SOURCE_DIR}/../../../../install/x64/Debug/geos/lib/geosd.lib" CACHE FILEPATH "")
set(GDAL_INCLUDE_DIR "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/gdal/include" CACHE PATH "")
set(GDAL_LIBRARY "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/gdal/lib/gdal_i.lib" CACHE FILEPATH "")
set(GDAL_LIBRARY_DEBUG "${CMAKE_SOURCE_DIR}/../../../../install/x64/Debug/gdal/lib/gdal_i.lib" CACHE FILEPATH "")
#set(SQLITE3_INCLUDE_DIR "C:/Library/install/sqlite3/include" CACHE FILEPATH "")
#set(SQLITE3_LIBRARY "C:/Library/install/sqlite3/lib/sqlite3.lib" CACHE FILEPATH "")
#set(V8_INCLUDE_DIR "C:/Library/install/v8/include" CACHE FILEPATH "")
#set(V8_LIBRARY "C:/Library/install/v8/lib/v8.lib" CACHE FILEPATH "")
#set(BUILD_OPENTHREADS_WITH_QT 1 CACHE BOOL "")
set(OSG_MSVC_VERSIONED_DLL 0 CACHE BOOL "")
set(CPACK_BINARY_NSIS 0 CACHE BOOL "")

set(LEVELDB_INCLUDE_DIR "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/leveldb/include" CACHE PATH "")
set(LEVELDB_LIBRARY "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/leveldb/lib/leveldb_static.lib" CACHE FILEPATH "")
set(LEVELDB_LIBRARY_DEBUG "${CMAKE_SOURCE_DIR}/../../../../install/x64/Debug/leveldb/lib/leveldb_static.lib" CACHE FILEPATH "")

# set(ENV{QTDIR} "C:/Qt/5.3/msvc2013_64_opengl" CACHE PATH "")
# set(DESIRED_QT_VERSION "5" CACHE STRING "")
# set(CMAKE_PREFIX_PATH "C:/Qt/5.3/msvc2013_64_opengl" CACHE PATH "")

set(SQLITE3_INCLUDE_DIR "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/sqlite3/include" CACHE PATH "")
set(SQLITE3_LIBRARY "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/sqlite3/lib/sqlite3.lib" CACHE FILEPATH "")
set(SQLITE3_LIBRARY_DEBUG "${CMAKE_SOURCE_DIR}/../../../../install/x64/Debug/sqlite3/lib/sqlite3.lib" CACHE FILEPATH "")

# set(Protobuf_INCLUDE_DIR "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/protobuf/include" CACHE PATH "")
# set(Protobuf_LIBRARY "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/protobuf/lib/libprotobuf.lib" CACHE PATH "")
# set(Protobuf_LIBRARY_DEBUG "${CMAKE_SOURCE_DIR}/../../../../install/x64/Debug/protobuf/lib/libprotobuf.lib" CACHE PATH "")
# set(Protobuf_PROTOC_EXECUTABLE "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/protobuf/bin/protoc.exe" CACHE PATH "")
# set(Protobuf_PROTOC_LIBRARY "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/protobuf/lib/libprotoc.lib" CACHE PATH "")
# set(Protobuf_PROTOC_LIBRARY_DEBUG "${CMAKE_SOURCE_DIR}/../../../../install/x64/Debug/protobuf/lib/libprotoc.lib" CACHE PATH "")
# set(Protobuf_LITE_LIBRARY "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/protobuf/lib/libprotobuf-lite.lib" CACHE PATH "")
# set(Protobuf_LITE_LIBRARY_DEBUG "${CMAKE_SOURCE_DIR}/../../../../install/x64/Debug/protobuf/lib/libprotobuf-lite.lib" CACHE PATH "")

set(POCO_INCLUDE_DIR "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/poco/include" CACHE PATH "")
set(POCO_FOUNDATION_LIBRARY "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/poco/lib/PocoFoundation.lib" CACHE PATH "")
set(POCO_FOUNDATION_LIBRARY_DEBUG "${CMAKE_SOURCE_DIR}/../../../../install/x64/Debug/poco/lib/PocoFoundationd.lib" CACHE PATH "")
set(POCO_NET_LIBRARY "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/poco/lib/PocoNet.lib" CACHE PATH "")
set(POCO_NET_LIBRARY_DEBUG "${CMAKE_SOURCE_DIR}/../../../../install/x64/Debug/poco/lib/PocoNetd.lib" CACHE PATH "")
set(POCO_UTIL_LIBRARY "${CMAKE_SOURCE_DIR}/../../../../install/x64/Release/poco/lib/PocoUtil.lib" CACHE PATH "")
set(POCO_UTIL_LIBRARY_DEBUG "${CMAKE_SOURCE_DIR}/../../../../install/x64/Debug/poco/lib/PocoUtild.lib" CACHE PATH "")

# set(SILVERLINING_DIR "D:/SilverLining SDK" CACHE PATH "")
# set(TRITON_DIR "D:/Triton SDK" CACHE PATH "")

set(OSGEARTH_INSTALL_SHADERS true CACHE BOOL "")
```

**Comment by Aea** [comment]
No idea, but I'm unsubscribing so you shouldn't be pinging me any more in this thread. First time I've been mentioned like this so I'll just assume it's a one-off fluke :)

**Comment by beachwalker** [comment]
I have found a solution, the Ws2_32.lib seems to be missing in the linked linbraries on Windows.

I have added TARGET_LINK_LIBRARIES(${LIB_NAME} Ws2_32.lib) to the WIN32 section in the osgearthUtil CMakeLists.txt as follows:

```
IF (WIN32)
  LINK_EXTERNAL(${LIB_NAME} ${TARGET_EXTERNAL_LIBRARIES} ${CMAKE_THREAD_LIBS_INIT} ${MATH_LIBRARY})
  TARGET_LINK_LIBRARIES(${LIB_NAME} Ws2_32.lib)
ELSE(WIN32)
  LINK_EXTERNAL(${LIB_NAME} ${TARGET_EXTERNAL_LIBRARIES} ${CMAKE_THREAD_LIBS_INIT} ${MATH_LIBRARY})
ENDIF(WIN32)
```

I think, this is just a quick fix and there is probably a better way to do following the coding standards of osgEarth. So I don't make a pull request and leave this open until someone from the devs takes a note and closes this with a better solution.
